cli/cmd/addon_download.go

```diff
@@ -0,0 +1,180 @@
+package cmd
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	cloudquery_api "github.com/cloudquery/cloudquery-api-go"
+	"github.com/cloudquery/cloudquery-api-go/auth"
+	"github.com/spf13/cobra"
+)
+
+const (
+	addonDownloadShort = "Download addon from CloudQuery Hub."
+	addonDownloadLong  = `Download addon from CloudQuery Hub.
+
+This downloads an addon from CloudQuery Hub to local disk.
+`
+	addonDownloadExample = `
+# Download an addon to local disk
+cloudquery addon download addon-team/addon-type/addon-name@v1.0.0`
+)
+
+func newCmdAddonDownload() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "download addon-team/addon-type/addon-name@v1.0.0 [-t directory]",
+		Short:   addonDownloadShort,
+		Long:    addonDownloadLong,
+		Example: addonDownloadExample,
+		Args:    cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// Set up a channel to listen for OS signals for graceful shutdown.
+			ctx, cancel := context.WithCancel(cmd.Context())
+
+			sigChan := make(chan os.Signal, 1)
+			signal.Notify(sigChan, syscall.SIGTERM)
+
+			go func() {
+				<-sigChan
+				cancel()
+			}()
+
+			return runAddonDownload(ctx, cmd, args)
+		},
+	}
+	cmd.Flags().StringP("target", "t", ".", `Download to specified directory. Use - for stdout`)
+
+	return cmd
+}
+
+func runAddonDownload(ctx context.Context, cmd *cobra.Command, args []string) error {
+	tc := auth.NewTokenClient()
+	token, err := tc.GetToken()
+	if err != nil {
+		return fmt.Errorf("failed to get auth token: %w", err)
+	}
+
+	addonParts := strings.Split(args[0], "/")
+	if len(addonParts) != 3 {
+		return fmt.Errorf("invalid addon ref: %s", args[0])
+	}
+	addonVer := strings.Split(addonParts[2], "@")
+	if len(addonVer) != 2 {
+		return fmt.Errorf("invalid addon ref %q: no version specified", args[0])
+	}
+	if !strings.HasPrefix(addonVer[1], "v") {
+		return fmt.Errorf("invalid addon ref %q: version must start with 'v'", args[0])
+	}
+
+	c, err := cloudquery_api.NewClientWithResponses(getEnvOrDefault(envAPIURL, defaultAPIURL),
+		cloudquery_api.WithRequestEditorFn(func(ctx context.Context, req *http.Request) error {
+			req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+			return nil
+		}))
+	if err != nil {
+		return fmt.Errorf("failed to create hub client: %w", err)
+	}
+
+	target, err := cmd.Flags().GetString("target")
+	if err != nil {
+		return err
+	}
+
+	addon, addonVersion, err := getAddonMetadata(ctx, c, addonParts[0], addonParts[1], addonVer[0], addonVer[1])
```

Comment:
We need to make two separate calls before attempting a download, because we don't know about the addon format (file extension) or this version's checksum beforehand.

Reply:
In the future we should look into amending the http headers (I'm not sure if it works alongside 302/Location or we might look into setting them in the bucket object) to carry the filename and expected checksum so we don't need to query these at all.